File: frontend/src/api/usage.ts
import client from '@/api/client'

/**
 * What the library has spent.
 *
 * Every figure here is a list-price estimate unless `estimated` says otherwise, and the
 * UI is required to say so. A cost from the pricing table is a published rate, not a
 * bill — providers change prices and offer discounts it does not model.
 */

interface DataResponse<T> {
  data: T
}

export interface UsageTotals {
  total_events: number
  /**
   * Fewer than `total_events` means some calls could not be priced — a custom endpoint,
   * or a model the table does not know — so `cost` is a floor rather than the whole
   * story.
   */
  priced_events: number
  cost: number
  currency: string
  estimated: boolean
  tokens: number
  seconds: number
}

export interface ProviderTotal {
  provider: string
  events: number
  units: number
  /** null when nothing in this group could be priced. */
  cost: number | null
}

export interface UsageSummary {
  totals: UsageTotals
  by_provider: ProviderTotal[]
}

export const usageApi = {
  summary(): Promise<UsageSummary> {
    return client.get<DataResponse<UsageSummary>>('/usage').then((r) => r.data.data)
  },

  forAsset(assetId: string): Promise<UsageTotals> {
    return client
      .get<DataResponse<UsageTotals>>(`/usage/assets/${assetId}`)
      .then((r) => r.data.data)
  },
}

/**
 * A cost as a string, at a precision that does not pretend.
 *
 * Enrichment costs land in fractions of a cent, and rounding those to two decimals shows
 * "$0.00" for work that was not free. Four decimals below a cent, two above.
 */
export function formatCost(cost: number, currency = 'USD'): string {
  const symbol = currency === 'USD' ? '$' : `${currency} `
  if (cost === 0) return `${symbol}0.00`
  if (cost < 0.01) return `${symbol}${cost.toFixed(4)}`
  return `${symbol}${cost.toFixed(2)}`
}

File: backend/app/routers/usage.py
"""What the library has spent, and on what.

FR 8.1.4. Every response here carries `estimated`, and the frontend is required to say
so: a figure from `usage/pricing.py` is a published list price, not a bill, and a number
shown without that qualification is a claim this app cannot support.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlmodel import Session

from app.auth import CurrentUser
from app.database import get_session
from app.models.asset import Asset
from app.schemas import DataResponse
from app.usage import events as usage_events

router = APIRouter()


class ProviderTotal(BaseModel):
    provider: str
    events: int
    units: int
    # None when nothing in this group could be priced — a `custom` endpoint, or a model
    # family the table does not know.
    cost: float | None = None


class UsageTotals(BaseModel):
    total_events: int
    # Fewer than `total_events` means some calls could not be priced, and the total below
    # is therefore a floor rather than the whole story. The UI says so.
    priced_events: int
    cost: float
    currency: str
    estimated: bool
    tokens: int
    seconds: int


class UsageSummary(BaseModel):
    totals: UsageTotals
    by_provider: list[ProviderTotal]


@router.get("", response_model=DataResponse[UsageSummary])
def read_usage(
    user: CurrentUser, session: Session = Depends(get_session)
) -> DataResponse[UsageSummary]:
    return DataResponse(
        data=UsageSummary(
            totals=UsageTotals(**usage_events.totals_for(session, user.id)),
            by_provider=[
                ProviderTotal(**row) for row in usage_events.by_provider(session, user.id)
            ],
        )
    )


@router.get("/assets/{asset_id}", response_model=DataResponse[UsageTotals])
def read_asset_usage(
    asset_id: str,
    user: CurrentUser,
    session: Session = Depends(get_session),
) -> DataResponse[UsageTotals]:
    """What has been spent on one asset.

    The asset is checked rather than trusted, even though a usage row carries its own
    user_id: an id that is not yours should not be a way to learn that it exists.
    """
    asset = session.get(Asset, asset_id)
    if not asset or asset.user_id != user.id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "not_found", "message": "No such asset"},
        )

    return DataResponse(data=UsageTotals(**usage_events.totals_for(session, user.id, asset_id)))
